File: 2025-12-01/part_1.rb
#!/usr/bin/env ruby

require "stringio"

# https://adventofcode.com/2025/day/1

def solve(input)
  pointer = 50
  zeros = 0

  while rotation = input.readline&.chomp
    dir, dist = rotation.split(/(\d+)/)
    pointer += case dir
               when 'L' then -dist.to_i
               when 'R' then dist.to_i
               else
                 raise "Invalid direction: #{dir}"
               end

    pointer %= 100
    zeros += 1 if pointer == 0

    break if input.eof?
  end

  zeros
end

# Tests ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

testfile = StringIO.new(<<~INPUT)
  L68
  L30
  R48
  L5
  R60
  L55
  L1
  L99
  R14
  L82
INPUT

tests = {
  testfile => 3
}.map do |(input, expectation)|
  result = solve(input)
  [input, [result == expectation, "Expected #{result} to equal #{expectation}"]]
end.to_h

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

if tests.values.all? { |v| v[0] }
  File.open(File.join(__dir__, 'input.txt'), 'r') do |file|
    puts solve(file)
  end
else
  $stderr.puts tests
end

File: 2025-12-01/part_1.swift
// https://adventofcode.com/2025/day/1

import Foundation

enum Failure: Error {
    case message(String)
}

func solve(url: URL) async throws -> Int {
    let handle = try FileHandle(forReadingFrom: url)
    defer { try? handle.close() }

    var zeros = 0
    var pointer = 50

    for try await line in handle.bytes.lines {
        guard let dist = Int(line.replacing("L", with: "-").replacing("R", with: "")) else {
            throw Failure.message("dist could not be converted")
        }

        pointer = (pointer + dist) % 100

        if pointer == 0 {
            zeros += 1
        }
    }

    return zeros
}

let file = URL(fileURLWithPath: #filePath)
let dir  = file.deletingLastPathComponent()
let target = dir.appendingPathComponent("input.txt")

let result = try await solve(url: target)
print(result)
